br/pkg/task/restore.go

```diff
@@ -1376,35 +1413,162 @@ func dropToBlackhole(
+				continue
+			}
+			// Add table to tableMap using table ID as key
+			tableMap[table.Info.ID] = table
+
+			// Add files to fileMap using file name as key
+			for _, file := range table.Files {
+				fileMap[file.Name] = file
+			}
+		}
+	}
+
+	// sanity check
+	if len(dbMap) == 0 && len(tableMap) != 0 {
+		err = errors.Annotate(berrors.ErrRestoreInvalidBackup, "contains tables but no databases")
+	}
+	return
+}
+
+func adjustTablesToRestoreAndCreateTableTracker(
+	logBackupTableHistory *stream.LogBackupTableHistoryManager,
+	cfg *RestoreConfig,
+	client *snapclient.SnapClient,
+	fileMap map[string]*backuppb.File,
+	tableMap map[int64]*metautil.Table,
+) (err error) {
+	snapshotDBMap := client.GetDatabaseMap()
+
+	// build tracker for pitr restore to use later
+	piTRTableTracker := utils.NewPiTRTableTracker()
+
+	// put all the newly created db that matches the filter during log backup into the pitr filter
+	newlyCreatedDBs := logBackupTableHistory.GetNewlyCreatedDBHistory()
+	for dbId, dbName := range newlyCreatedDBs {
+		if utils.MatchSchema(cfg.TableFilter, dbName) {
+			piTRTableTracker.AddDB(dbId)
+		}
+	}
+
+	// get all the tables seen during the log backup
+	tableHistory := logBackupTableHistory.GetTableHistory()
+
+	for tableID, dbIDAndTableName := range tableHistory {
+		start := dbIDAndTableName[0]
+		end := dbIDAndTableName[1]
+
+		var dbName string
+		// check in snapshot
+		if snapDb, exists := snapshotDBMap[end.DbID]; exists {
+			dbName = snapDb.Info.Name.O
+		} else if name, exists := logBackupTableHistory.GetDBNameByID(end.DbID); exists {
+			// check during log backup
+			dbName = name
+		} else {
+			log.Warn("did not find db id in full/log backup, "+
+				"likely the full&log backup provided have different filters, ignoring this db",
+				zap.Any("dbId", end.DbID))
+			continue
+		}
+
+		// skip if db doesn't match
+		if !utils.MatchSchema(cfg.TableFilter, dbName) {
+			continue
+		}
+
+		// handle in filter range cases
+		// 1. original == current, didn't have renaming
+		// 2. original has been renamed and current is in the filter range
+		// we need to restore original table
+		if utils.MatchTable(cfg.TableFilter, dbName, end.TableName) {
+			// put this db/table id into pitr tracker as it matches with user's filter
+			// have to update filter here since table might be empty or not in snapshot so nothing will be returned .
+			// but we still need to capture this table id to restore during log restore.
+			piTRTableTracker.AddTable(end.DbID, tableID)
+
+			// check if snapshot contains the original db/table
+			originalDB, exists := snapshotDBMap[start.DbID]
+			if !exists {
+				// original db created during log backup, snapshot doesn't have information about this db so doesn't
+				// need to restore at snapshot
+				continue
+			}
+
+			// need to restore the matching table in snapshot restore phase
+			for _, originalTable := range originalDB.Tables {
```

Comment:
I prefer to merge `filterRestoreFiles` and `adjustTablesToRestoreAndCreateTableTracker` without any intermediates. That's because I am worried about the time complexity of the loop.

Reply:
good question, discussed offline, the case of rename/exchange table should be rare so it should be fine, if it's problem we can convert it to be a map in future.